examples/vae.py

```diff
@@ -0,0 +1,110 @@
+from __future__ import absolute_import, division, print_function
+
+import argparse
+import os
+from collections import OrderedDict
+
+import torch
+import torch.utils.data
+from torch import nn, optim
+from torch.nn import functional as F
+from torchvision import datasets, transforms
+
+import funsor
+import funsor.distributions as dist
+import funsor.ops as ops
+from funsor.domains import bint, reals
+
+REPO_PATH = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+DATA_PATH = os.path.join(REPO_PATH, 'data')
+
+
+class Encoder(nn.Module):
+    def __init__(self):
+        super(Encoder, self).__init__()
+        self.fc1 = nn.Linear(784, 400)
+        self.fc21 = nn.Linear(400, 20)
+        self.fc22 = nn.Linear(400, 20)
+
+    def forward(self, image):
+        image = image.reshape(image.shape[:-2] + (-1,))
+        h1 = F.relu(self.fc1(image))
+        loc = self.fc21(h1)
+        scale = self.fc22(h1).exp()
+        return loc, scale
+
+
+class Decoder(nn.Module):
+    def __init__(self):
+        super(Decoder, self).__init__()
+        self.fc3 = nn.Linear(20, 400)
+        self.fc4 = nn.Linear(400, 784)
+
+    def forward(self, z):
+        h3 = F.relu(self.fc3(z))
+        out = torch.sigmoid(self.fc4(h3))
+        return out.reshape(out.shape[:-1] + (28, 28))
+
+
+def main(args):
+    encoder = Encoder()
+    decoder = Decoder()
+
+    encode = funsor.torch.function(reals(28, 28), (reals(20), reals(20)))(encoder)
+    decode = funsor.torch.function(reals(20), reals(28, 28))(decoder)
+
+    @funsor.interpreter.interpretation(funsor.montecarlo.monte_carlo)
+    def loss_function(data, subsample_scale):
+        # Lazily sample from the guide.
+        loc, scale = encode(data)
+        q = funsor.Independent(
+            dist.Normal(loc['i'], scale['i'], value='z'),
+            'z', 'i')
+
+        # Evaluate the model likelihood at the lazy value z.
+        probs = decode('z')
```

Comment:
nit: we might want to use a `Variable` object here (and in any similar situations) as it's not immediately clear that `decode` is a `Funsor`.  We can fix example idioms in followup PRs though.